src/app.ts

```diff
@@ -37,9 +50,15 @@ export function entity<T = unknown>(
     });
 }
 
-export function activity(functionName: string, options: ActivityOptions): void {
+export function activity(functionName: string, options: ActivityOptions): RegisteredActivity {
     azFuncApp.generic(functionName, {
         trigger: trigger.activity(),
         ...options,
     });
+
+    const result: RegisteredActivity = (input?: unknown): RegisteredActivityTask => {
```

Comment:
I think this has been suggested before, but rather than this:
```
yield mySubOrchestrator(input);
yield mySubOrchestrator(input).withRetry(retryOptions);
```

What do we think of this?
```
yield mySubOrchestrator.call(input);
yield mySubOrchestrator.callWithRetry(input);
```

Reply:
I believe this was one of the proposal that we discussed, and we agreed that the former was cleaner. Part of the reason is, for the second option, the `callWithRetry()` would actually need to look something like:

```TS
yield mySubOrchestrator.callWithRetry({ input: 'someInput', retryOptions });
```

Otherwise, we would end up with the null/undefined problem, in case a user wants to call an activity/subOrchestrator with retry, but with no input:

```TS
yield mySubOrchestrator.callWithRetry(null, retryOptions);
```

Reply:
I think I was assuming the `retryOptions` argument was before `input`. Correct me if I'm wrong, but I think that's what we're [doing today](https://github.com/Azure/azure-functions-durable-js/blob/v3.x/types/orchestration.d.ts#L191). So in that case you'd never be forced to pass `null`

> If we go with an options object for .callWithRetry(), that would also beg the question of whether we should also have an options object for .call(), for consistency's sake

You bring up an interesting point. Won't we need an options object for `call` anyways? I believe this is what Jacob [said he was doing](https://github.com/Azure/azure-functions-durable-js/issues/418#issuecomment-1405387203) for C#, in part for things like `instanceId` and `fireAndForget` as well as the retry stuff. If we do an options object, we debatably don't need any `WithRetry` methods at all

Reply:
> I think I was assuming the retryOptions argument was before input. Correct me if I'm wrong, but I think that's what we're [doing today](https://github.com/Azure/azure-functions-durable-js/blob/v3.x/types/orchestration.d.ts#L191). So in that case you'd never be forced to pass null

Yes, that is a good point. 

> Won't we need an options object for call anyways?

Given where the node SDK is now, I'm not convinced we would need to add a fire-and-forget operation to calling activities and sub-orchestrators any time soon. I would still also prefer to have to explicitly state that you'd like to call an activity/suborchestrator with retry, rather than that being implied by an options object parameter. I remember us having similar conversations too around changing `context.df.callActivity()` and `context.df.callSubOrchestrator()` to accepting options object, and that was the same conclusion

Would like to hear @davidmrdavid's thoughts here too

Reply:
> Given where the node SDK is now, I'm not convinced we would need to add a fire-and-forget operation to calling activities and sub-orchestrators any time soon.

There are no fire-and-forget activities but the DF extension does support fire-and-forget sub-orchestrators. But that has not been a top ask for us at this time. Still, worth noting.

>  I remember us having similar conversations too around changing context.df.callActivity() and context.df.callSubOrchestrator() to accepting options object, and that was the same conclusion

In general, I think @hossam-nasr's design in this PR is giving the input to activities and sub-orchestrators special treatment so that:
(1) if a user has an input but not retryOptions, they don't have to construct a whole options object for it. I think this is a common scenario worth optimizing for.
(2) if a user has no input but they have `retryOptions`, they don't have to pass in a `null` input, which looks a bit clunky.

I like this design because I think inputs need to be handled separately from retry configuration as well as other knobs, as the use of the former is more common than the latter _afaik._

If we wanted to tweak the current design to allow for fire-and-forget sub-orchestrators, I think we could make the API look like this: